fix: adjust credential chain

ekristen opened the pull request:
<!--  Thanks for sending a pull request!  Before submitting:

1. Read our CONTRIBUTING.md guide
2. Rebase your PR if it gets out of sync with main
-->

PR #2889 was suppose to fix this but it doesn't as far as I can tell. It's possibly because I have IDMS enabled too and not blocked, but you can't expect that to get blocked to make things work. (I read about this in one of the issues)

Simply switching the chain flow up a bit resolves this issue without the need of if statements makes this all work again, which is what this PR does.

I've build the image and am hosting it at `ghcr.io/ekristen/tempo:2.2.3-ek1` if you'd like to test it out.

**What this PR does**:

Fixes AWS authentication for IRSA.

**Which issue(s) this PR fixes**:
Fixes #2888


**Checklist**
- [ ] Tests updated
- [x] Documentation removed
- [ ] `CHANGELOG.md` updated - the order of entries should be `[CHANGE]`, `[FEATURE]`, `[ENHANCEMENT]`, `[BUGFIX]`

## Configuring IRSA Testing with GitHub Actions

1. Identify a test AWS Account (ideally a dedicated account for this with nothing else)
2. Add GitHub Actions Tokens as an Identity Provider
  - url: https://token.actions.githubusercontent.com
  - audience: sts.amazonaws.com
3. Create an AWS Role named `tempo-tests`
4. Set the Trust Policy to the trust policy section
  - [ ] Replace 123456123456 with the real account id
  - [ ] Edit line 162 and replace 123456123456 with the real account id
6. Add NO permissions to the role. None. This will allow the role to be assumed but it'll have no permissions to do anything in the account. 
7. Complete

### Trust Policy

**Note:** replace 123456123456 with your account id.

```json
{
    "Version": "2012-10-17",
    "Statement": [
        {
            "Effect": "Allow",
            "Principal": {
                "Federated": "arn:aws:iam::123456123456:oidc-provider/token.actions.githubusercontent.com"
            },
            "Action": "sts:AssumeRoleWithWebIdentity",
            "Condition": {
                "StringLike": {
                    "token.actions.githubusercontent.com:sub": "repo:grafana/tempo:*"
                },
                "StringEquals": {
                    "token.actions.githubusercontent.com:aud": "sts.amazonaws.com"
                }
            }
        }
    ]
}
```

CLAassistant commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/not_signed)](https://cla-assistant.io/grafana/tempo?pullRequest=2958) <br/>Thank you for your submission! We really appreciate it. Like many open source projects, we ask that you sign our [Contributor License Agreement](https://cla-assistant.io/grafana/tempo?pullRequest=2958) before we can accept your contribution.<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla-assistant.io/check/grafana/tempo?pullRequest=2958) it.</sub>

mapno commented on tempodb/backend/s3/s3.go:
```diff
@@ -429,34 +429,28 @@ func createCore(cfg *Config, hedge bool) (*minio.Core, error) {
-				},
-			}),
-		}
+	chain := []credentials.Provider{
```
This chain seems to exclude the AWS SDK auth that was introduced in https://github.com/grafana/tempo/pull/2760. Is it not needed to fix the problem described in https://github.com/grafana/tempo/issues/2743?

ekristen replied:
It doesn't. That's what this chain does, since v6.0.53 of minio package. https://github.com/grafana/tempo/pull/2958/files#diff-c0f785bb00d262f48124144ac01da108c80587ee0cbadb13f4451942be17dab3R443-R447

The other code doesn't seem to work right. I didn't troubleshoot it in depth, except for IRSA doesn't work, the minio client sdk can handle all the various methods.

coufalja replied:
If that is removed the IDMSv1 will break again, I in parallel engaged with Minio and uncovered that it is actually a minio-go lib issue fixed by https://github.com/minio/minio-go/pull/1877 released in `v7.0.63`, if we could bump minio-go version we could revert the #2760.

knylander-grafana reviewed (commented):
Thank you for updating the documentation.

hobbsh commented:
Tested and it seems to solve our IRSA issue in EKS but I did not test any other cases.

ekristen commented:
All the other methods are pretty easy to test. I've tested this chain in another tool, happy to have whoever test to get it merged.

mapno commented:
> All the other methods are pretty easy to test.

How would you test them? We're looking for a reliable way to test all the different auth modes. We manually _tested_ the current implementation and it worked fine before.

 It feels we're going in circles with these S3 auth changes, as this PR is actually configuring the credential provider chain how it was in the first v2.2 release, before issue #2743 was reported and fixed.

ekristen commented:
@mapno The latest commit adds tests for the entire credential chain with calls to metadata service for instance profile role and IRSA being mocked. I excluded calls to ECS as I don't think that's a valid deployment target at this time.

Tests will start to fail due to GitHub Actions OIDC to an AWS account not being setup. These instructions are being put at the top of the PR to be followed as well.

1. Identify a test AWS Account (ideally a dedicated account for this with nothing else)
2. Add GitHub Actions Tokens as an Identity Provider
  - url: https://token.actions.githubusercontent.com
  - audience: sts.amazonaws.com
3. Create an AWS Role named `tempo-tests`
4. Set the Trust Policy to the trust policy section
  - [ ] Replace 123456123456 with the real account id
  - [ ] Edit line 162 and replace 123456123456 with the real account id
6. Add NO permissions to the role. None. This will allow the role to be assumed but it'll have no permissions to do anything in the account. 
7. Complete

### Trust Policy

**Note:** replace 123456123456 with your account id.

```json
{
    "Version": "2012-10-17",
    "Statement": [
        {
            "Effect": "Allow",
            "Principal": {
                "Federated": "arn:aws:iam::123456123456:oidc-provider/token.actions.githubusercontent.com"
            },
            "Action": "sts:AssumeRoleWithWebIdentity",
            "Condition": {
                "StringLike": {
                    "token.actions.githubusercontent.com:sub": "repo:grafana/tempo:*"
                },
                "StringEquals": {
                    "token.actions.githubusercontent.com:aud": "sts.amazonaws.com"
                }
            }
        }
    ]
}
```

mapno commented:
Wow, thanks for the tests. I'll need to take a detailed look at them.

> My suspicion to why it appear to not work in the first place is because one of the handlers in the chain probably retuned keys prior to getting to the IRSA chain.

But the chain barely changes from its current state to your changes, no? It's the [same chain](https://github.com/grafana/tempo/blob/main/tempodb/backend/s3/s3.go#L449-L459), except for `credentials.Static` that is put at the top of the chain. I run your test with the chain split like that and they passed as well. I don't understand what this PR is changing in that sense.

ekristen commented:
I think a lot of things got mixed together. 

It looks like there was some issues with IMDS and then IRSA (assume role with web identity)

If the minio library is updated to the latest based on the other individuals comment and other this is merged or the other PRs are reverted then things should be ok.

I updated the order of the chain to make more sense and added tests.

mapno commented:
Just reviewed the tests, they're a pretty nice addition! Do you think I'd be possible to mock the metadata service for IRSA with a tool like https://github.com/aws/amazon-ec2-metadata-mock, instead of using an actual AWS account? I don't think it's something we'll be able to do easily.

I still don't understand why the new chain works and the current doesn't, as it doesn't introduce changes from what we have now (or had in the past). Sorry if I'm being very obtuse here.

>If the minio library is updated to the latest based on the other individuals comment and other this is merged or the other PRs are reverted then things should be ok.

Would you like to take care of that change? I can do it if not.

Thanks for the patience! Love to see these contributions in Tempo ❤️

mapno commented:
I've opened a new PR based on this one, but with minio-go updated. There are images publised for [amd64](https://hub.docker.com/layers/mapno/tempo/aws-cred-fix-amd64/images/sha256-68104b88e948534457c8a5dff8c95799f3a8b24f3936589501c6c6a42604fc48?context=repo) and [arm64](https://hub.docker.com/layers/mapno/tempo/aws-cred-fix-arm64/images/sha256-5e513e3fa7cf840753b54721781a04e353cb1911f9076a046c3ffa634fd9f39f?context=repo) architectures. If all of you can test them, we can get that PR merged. cc/ @ekristen @coufalja @sbezr @z0rk @finda-yeongjo @hobbsh

ekristen commented:
Updated the dependency. I still thing this PR is the better way to go with the tests.

ekristen commented:
I will rebase.
